### ISSUE TITLE
Change folder hierarchy to split environments

### ISSUE THREAD

**Issue by morgante** [terraform-google-modules/terraform-example-foundation#47]
Currently, we use a "flat" hierarchy in each component where all resources for a component are contained in a single Terraform state. This increases the blast radius and makes it challenging to test changes in nonprod before productionizing.

I'd like to split this repo to have folders for each environment (under the components), like the structure described [here](https://docs.google.com/document/d/1lpuK7salw2gUPdpQByPFAS4IvmM2aeHE5eFx5W_WTd8/edit#heading=h.wkpi8bw1p19k):
```
├── foundations
│   ├── folders
│   │   └── shared
│   │       └── folders.tf
│   ├── logging
│   │   ├── nonprod
│   │   │   └── export.tf
│   │   └── prod
│   │       └── export.tf
│   ├── network
│   │   ├── nonprod
│   │   │   └── network.tf
│   │   └── prod
│   │       └── network.tf
│   ├── org-iam
│   │   └── shared
│   │       └── iam.tf
│   ├── org-policies
│   │   └── shared
│   │       └── org-policies.tf
│   ├── Jenkinsfile
│   ├── backend.tf
│   ├── common-providers.tf
│   ├── common-variables.tf
│   ├── main.tf
│   └── output.tf
```

**Comment by morgante** [comment]
@rjerrems @ericyz What are your thoughts here?

**Comment by rjerrems** [comment]
Yeah I agree with this in principle, we may also need to track some work on the bootstrap repo to change the way the triggers work to something more path based - or support that as an option? A couple of questions:

- Are you also proposing we removing the current defined numbered layers 1-org, 2-networks etc?
- Will the modules for each of those ^ still live in those layers or be in a central location?

**Comment by morgante** [comment]
> Yeah I agree with this in principle, we may also need to track some work on the bootstrap repo to change the way the triggers work to something more path based - or support that as an option?

Yes, I think we'll have to do some work there.

> Are you also proposing we removing the current defined numbered layers 1-org, 2-networks etc?

No, we should still keep the top-level layers but just add the environment folders underneath them

> Will the modules for each of those ^ still live in those layers or be in a central location?

For now, I think we should just add them into this repo and assume they'll be pushed to a central location.

**Comment by rjerrems** [comment]
Yeah okay cool, I am onboard if we keep the same top level folder structure -  it is a good idea to make it easier to roll out changes between environments, although we may still have some things that cut across the organization.

**Comment by xingao267** [comment]
Our team sort of seeing a similar question around cicd and using cloud build. If prod and non-prod shares certain common .tf files. How would you make sure changes made to those files only trigger non-prod job, not both non-prod and prod using cloud build?

@rjerrems FYI, we also have a [CICD module](https://github.com/GoogleCloudPlatform/healthcare-data-protection-suite/tree/master/templates/tfengine/components/cicd), or we call it a "component", which is **GitHub** Cloud Build trigger based, and also supports CSR.

**Comment by morgante** [comment]
> If prod and non-prod shares certain common .tf files. How would you make sure changes made to those files only trigger non-prod job, not both non-prod and prod using cloud build?

My recommended approach is via branching, essentially 1 branch per environment and changes are only applied to an env once you merge to that env branch.

I explain it more in this video: https://www.youtube.com/watch?v=3vfXQxWJazM (ping me for slides)

**Comment by mikelaramie** [comment]
I took a stab at this here:  https://github.com/mikelaramie/terraform-example-foundation/tree/feature-org-repo , but it looks like I may have botched something since it says it can't auto-merge to upstream (but can to my local fork)

**Comment by morgante** [comment]
@mikelaramie You probably just have merge conflicts that need to be resolved.

**Comment by bharathkkb** [comment]
closed via #51
